doc/paper/paper.tex

```diff
@@ -0,0 +1,1012 @@
+
+    In the SEQUIN language for the SEQ \cite{seq} database, an as-of join is expressed in
+    this way:
+
+    \begin{lstlisting}
+    // first define the moving average as a view
+    CREATE VIEW MovAvgStockl AS (
+    PROJECT avg(C.high) as avghigh
+    FROM stock1 c
+    OVER \$P-23 TO \$p.);
+
+    // then use the view in the query
+    PROJECT A.avghigh - B.high
+    FROM MovAvgStockl A,
+    Previous(PROJECT D.high
+    FROM Stock2 D
+    WHERE D.volume > 25,000) B
+    WHERE.\$P > 2000;
+    \end{lstlisting}
+
+    The \textit{Previous} function returns the most recent hour at a point during
+    evaluation when the volume on Stock2 exceeded 25,000. In dp3 we have no support
+    for aggregations at this time, so we are unable to compute the moving average
+    represented in the view. In dp3, \q{Previous} is implemented with \q{precedes
+    immediate}.
+
+    \subsection{Work remaining}
+    Significant work remains, both in design validation and feature development, to
+    get dp3 to a useful state.
+
+    \subsubsection{Design validation}
+    The largest outstanding design question relates to dp3’s choice to use MCAP
+    as the data format in leaf storage. There are a few benefits to this
+    choice: (1) No transcoding required on ingestion or query. (2)
+    Automatically compatible with user’s own internal tooling. (3) Streaming
+    playback is a row-oriented workload. (4) Automatically compatible with
+    common robotics recording workflows
+
+    There are also drawbacks, primarily in the performance department. (1)
+    Columnar storage would realize better compression in general. (2) A
+    standard columnar format may allow us to leverage an off-the-shelf query
+    executor, and could probably support more performant queries for our search
+    workload. (3) Building generic systems on MCAP requires supporting query
+    operations on multiple different binary encoding formats in place, which is
+    tedious to implement and support, and may carry some performance risks.
+
+    The question is also complicated by workload-specific factors. If a
+    columnar format is used, it is likely that small \q{row groups} or whatever
+    equivalent concept exists will be required, to enable queries to merge
+    large numbers of tables at once without using huge amounts of memory. Using
+    small row groups will tend to work against the interests of good OLAP
+    performance on the data files, obviating some of the benefits we would be
+    hoping to get with the columnar format.
+
+    Furthermore, a significant portion of the heavy read volume the database will
+    serve for ML purposes, will be image and point-cloud data. In these kinds of
+    messages, there is effectively one column: a giant byte array. This means that
+    row-oriented and columnar files for these kinds of data are physically pretty
+    similar. So intuitively, it seems like the performance benefits we would hope
```

Comment:
"Intuitively," "it seems," "we would hope," "mostly," etc. are subjective claims that are especially weak when paired together (as in this sentence). Can we omit or more clearly describe/quantify these?